vcd/provider.go

```diff
@@ -108,6 +109,33 @@ func Provider() terraform.ResourceProvider {
 	}
 }
 
+// This is a global MutexKV for all resources
+var vcdMutexKV = mutexkv.NewMutexKV()
+
+func lockVapp(d *schema.ResourceData) {
+	vcdMutexKV.Lock(d.Get("vdc").(string) + d.Get("name").(string))
```

Comment:
There should be a separator character between names to minimize the risk of collision. For example, VDC "pant" and entity "ear" generates the same lock name as "pan" + "tear". If we create the names as "pant#ear" and "pan#tear", we avoid this situation. While this case is unlikely to occur in production, it is possible during testing, when users come up with not-well-thought short names.

Reply:
Good catch! I'd like to see the separator for another reason - when printing an error/debug message it will be a lot easier to read when there's a separator. I just hope that no-one is using the hash sign in the names, like "Obj \#1", "Obj \#2"... Maybe let's use comma as in "pan,tear" or pipe "pan|tear"?

Reply:
> I just hope that no-one is using the hash sign in the names

We are. In function `(c *Config) Client()` (`config.go`)

Reply:
That's a bit different though (used for connection cache pointers). What I meant is that you can name your VDC like "Test VDC #1" or vApp "My vApp #2" and the hash separator would result in a crash.

![Screen Shot 2019-06-03 at 10 38 11](https://user-images.githubusercontent.com/11330689/58784337-cd51a700-85eb-11e9-9e3e-b9fd0d62b38e.png)

Reply:
I misunderstood the remark. Using commas (or dashes, colons, or pipe signs)  is fine

Reply:
Will use PIPE

Reply:
Changed